### ISSUE TITLE
TCP sockets

### ISSUE THREAD

**Issue by streamich** [uNetworking/uWebSockets.js#370]
This project currently supports WebSockets and HTTP server. Would you consider also exposing hooks for users to work with TCP sockets? The API would be somewhat similar to that of WebSockets with the required differences. Idea being that users who want high performance EPOLL solution could use this lib instead of `libuv` and Node.js users could have something faster than [Node.js `net` module](https://nodejs.org/api/net.html).

**Comment by alexhultman** [comment]
Controversial topic - this request/subject has surfaced a few times before.

* uWS is a **high level** library for doing **high performance** web messaging. That is the point and purpose of this project.

It is high level because when you do as much as possible in native C++ and only let the very **business logic** be scripted, you have minimal performance loss and maximum productivity.

You get a sweet point where you can, with 4 lines of JavaScript, set up a WebSocket server that outperforms most of all alternatives from within Node.js. That is the reason why this project is what it is.

I cannot think of one single case where **business logic** in a web project would need to directly use TCP or EPOLL.

**Comment by streamich** [comment]
@alexhultman what if your business logic is to write a serivce for IoT devices which communicates with those devices using MQTT, STOMP and other protocols? Yes, it would be nice to also have a parser for those protocols in C++, but we are not going to add parsers for all protocols in this project.

Say you wanted to implement MQTT using uWS, at least for my use cases, the only thing that could be moved to C++ would be the MQTT packet encoder, but until we don't have that I would be happy to use Node.js MQTT parser with uWS as EPOLL.

**Comment by streamich** [comment]
Another use case would be gRPC. (AWS ALB recently added support for gRPC.)

**Comment by streamich** [comment]
If uWS TCP stack is faster than Node, another HUGE use case is implementing all database drivers (Redis, MySQL, PostgreSQL, etc..) using uWS for TCP.

**Comment by alexhultman-2fa** [comment]
> I mean, if you could have a database driver which is 2x faster, everyone would want that.

I have presented initial code for a database driver that is 12x faster. Nobody cared and I have no interest in developing yet another thing that really only shows what we all know already: that scripting languages are not about performance. This is computer science 101.

I mean I could rewrite everything that Node.js does, in C++, and have incredible performance. But that would really just be a C++ server, not Node.js. Node.js is Node.js because everyone and their mother knows JavaScript and insist on using it everywhere for anything.

With a database the bottleneck is (mostly) going to be in the database itself. Sure, fast in-memory databases will bottleneck in the connector if done bad (and Node.js has really bad connectors), but it is still not a project I want to take on.

Because, none of this is news to anyone who knows computer science. It's all about popularity.

TLDR; This project is about making a statement, not to polish a turd into slightly less of a turd. You either do something really good, or you don't. This project is about tackling the web messaging problematics, really good.

Besides, the TCP support of uSockets is highly dependent on C-pointers and memory alignments that you don't have in JS - so it would have to be some kind of shoe-horned wrapper turd.

**Comment by alexhultman-2fa** [comment]
This project is about presenting **simple solutions** to **business problematics**, with **very competitive** results.
